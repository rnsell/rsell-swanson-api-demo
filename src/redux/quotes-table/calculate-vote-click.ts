import { match } from "ts-pattern";
import { ClickOnQuote, ButtonClickType } from "./quotes-table.actions";
import { Quote } from "./quotes-table.model";

const downVoteClick = {
  payload: {
    clickType: ButtonClickType.DOWN_VOTE,
  },
};

const upVoteClick = {
  payload: {
    clickType: ButtonClickType.UP_VOTE,
  },
};

// id: string;
// quoteText: string;
// votes: number;
// upVoteSelected: boolean;
// downVoteSelected: boolean;

export const calculateNewVoteClick = (
  quote: Quote,
  clickQuoteAction: ClickOnQuote
): Quote => {
  return (
    match<[Quote, ClickOnQuote], Quote>([quote, clickQuoteAction])
      // Univted
      .with(
        [{ upVoteSelected: false, downVoteSelected: true }, upVoteClick],
        ([oldQuote]) => {
          const upVoteSelected = true;
          const downVoteSelected = false;
          const votes = (oldQuote.votes += 2);
          return { ...oldQuote, upVoteSelected, downVoteSelected, votes };
        }
      )
      .with(
        [{ upVoteSelected: false, downVoteSelected: false }, upVoteClick],
        ([oldQuote]) => {
          const upVoteSelected = true;
          const downVoteSelected = false;
          const votes = (oldQuote.votes += 1);
          return { ...oldQuote, upVoteSelected, downVoteSelected, votes };
        }
      )
      .with(
        [{ upVoteSelected: true, downVoteSelected: false }, upVoteClick],
        ([oldQuote]) => {
          const upVoteSelected = false;
          const downVoteSelected = false;
          const votes = (oldQuote.votes -= 1);
          return { ...oldQuote, upVoteSelected, downVoteSelected, votes };
        }
      )
      .with(
        [{ upVoteSelected: false, downVoteSelected: false }, downVoteClick],
        ([oldQuote]) => {
          const upVoteSelected = false;
          const downVoteSelected = true;
          const votes = (oldQuote.votes -= 1);
          return { ...oldQuote, upVoteSelected, downVoteSelected, votes };
        }
      )
      .with(
        [{ upVoteSelected: false, downVoteSelected: true }, downVoteClick],
        ([oldQuote]) => {
          const upVoteSelected = false;
          const downVoteSelected = false;
          const votes = (oldQuote.votes += 1);
          return { ...oldQuote, upVoteSelected, downVoteSelected, votes };
        }
      )
      .with(
        [{ upVoteSelected: true, downVoteSelected: false }, downVoteClick],
        ([oldQuote]) => {
          const upVoteSelected = false;
          const downVoteSelected = true;
          const votes = (oldQuote.votes -= 2);
          return { ...oldQuote, upVoteSelected, downVoteSelected, votes };
        }
      )
      .otherwise(() => quote)
  );
};
