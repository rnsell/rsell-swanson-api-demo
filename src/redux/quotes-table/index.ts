export * from "./quotes-table.actions";
export * from "./quotes-table.model";
export * from "./quotes-table.reducers";
