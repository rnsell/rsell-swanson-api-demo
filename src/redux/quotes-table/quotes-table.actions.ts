import { createAction } from "@reduxjs/toolkit";

export type PopulateQuotesAction = {
  type: "QUOTES_TABLE::POPULATE";
  payload: Array<string>;
};

export enum ButtonClickType {
  UP_VOTE = "UP_VOTE",
  DOWN_VOTE = "DOWN_VOTE",
}

export type ClickOnQuote = {
  type: "QUOTES_TABLE::CLICK";
  payload: {
    clickType: ButtonClickType;
    quoteId: string;
  };
};

export type QuotesTableActions = PopulateQuotesAction | ClickOnQuote;

export const populateQuotesTable = createAction<Array<string>>(
  "QUOTES_TABLE::POPULATE"
);

export const upVote = (quoteId: string): ClickOnQuote => {
  return {
    type: "QUOTES_TABLE::CLICK",
    payload: {
      clickType: ButtonClickType.UP_VOTE,
      quoteId,
    },
  };
};

export const downVote = (quoteId: string): ClickOnQuote => {
  return {
    type: "QUOTES_TABLE::CLICK",
    payload: {
      clickType: ButtonClickType.DOWN_VOTE,
      quoteId,
    },
  };
};
