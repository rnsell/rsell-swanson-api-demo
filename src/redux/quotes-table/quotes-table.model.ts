import hash from "object-hash";
import { fromPairs, random } from "lodash";

export class Quote {
  id: string;
  quoteText: string;
  votes: number;
  upVoteSelected: boolean;
  downVoteSelected: boolean;

  constructor(data: {
    quoteText: string;
    votes?: number;
    id?: string;
    upVoteSelected?: boolean;
    downVoteSelected?: boolean;
  }) {
    const { quoteText, votes, id, upVoteSelected, downVoteSelected } = data;
    this.id = id ?? hash(quoteText, { algorithm: "md5" });
    this.quoteText = quoteText;
    // Lets simulate some votes already existed
    this.votes = votes ?? random(-2, 3);
    this.upVoteSelected = upVoteSelected ?? false;
    this.downVoteSelected = downVoteSelected ?? false;
  }
}

export type QuotesTable = {
  [key: string]: Quote;
};

export const generateQuotesTable = (quotes: string[]): QuotesTable => {
  const table = fromPairs(
    quotes.map((quoteText) => {
      const newQuote = new Quote({ quoteText });
      return [newQuote.id, newQuote];
    })
  );

  return table;
};
