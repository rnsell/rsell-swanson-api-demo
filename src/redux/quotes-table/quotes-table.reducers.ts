import { match } from "ts-pattern";
import { QuotesTableActions } from "./quotes-table.actions";
import { QuotesTable } from "./quotes-table.model";
import {
  handlePopulateTable,
  handleQuotesTableVoteClick,
} from "./reducer-handlers";

const defaultQuoteTable: QuotesTable = {};

export const quoteTableReducer = (
  state: QuotesTable = defaultQuoteTable,
  action: QuotesTableActions
): QuotesTable =>
  match<[QuotesTableActions], QuotesTable>([action])
    .with([{ type: "QUOTES_TABLE::POPULATE" }], handlePopulateTable)
    .with([{ type: "QUOTES_TABLE::CLICK" }], handleQuotesTableVoteClick(state))
    .otherwise(() => state);
