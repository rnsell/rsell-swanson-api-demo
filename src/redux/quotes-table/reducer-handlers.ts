import { calculateNewVoteClick } from "./calculate-vote-click";
import { ClickOnQuote, PopulateQuotesAction } from "./quotes-table.actions";
import { generateQuotesTable, QuotesTable } from "./quotes-table.model";

export const handlePopulateTable = ([action]: [PopulateQuotesAction]) => {
  const { payload: quotes } = action;
  const newQuotesTable = generateQuotesTable(quotes);

  return newQuotesTable;
};

export const handleQuotesTableVoteClick = (state: QuotesTable) => ([action]: [
  ClickOnQuote
]): QuotesTable => {
  const { payload } = action;
  const { quoteId } = payload;
  const quote = state[quoteId];

  if (!quote) {
    return state;
  }

  const newQuote = calculateNewVoteClick(quote, action);
  const quoteRecord = { [newQuote.id]: newQuote };

  return { ...state, ...quoteRecord };
};
