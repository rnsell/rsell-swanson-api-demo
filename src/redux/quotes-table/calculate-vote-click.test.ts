import { calculateNewVoteClick } from "./calculate-vote-click";
import { downVote, upVote } from "./quotes-table.actions";
import { Quote } from "./quotes-table.model";

const baseQuote = new Quote({
  quoteText: "Go Beyond Plus Ultra",
  upVoteSelected: false,
  downVoteSelected: false,
  votes: 0,
});

const upVotedQuote = new Quote({
  quoteText: "Go Beyond Plus Ultra",
  upVoteSelected: true,
  downVoteSelected: false,
  votes: 0,
});

const downVotedQuote = new Quote({
  quoteText: "Go Beyond Plus Ultra",
  upVoteSelected: false,
  downVoteSelected: true,
  votes: 0,
});

const upVoteClickAction = upVote("someId");
const downVoteClickAction = downVote("someId");

describe("calculate-vote-click", () => {
  it("upvote cases", () => {
    const votesCases = [
      calculateNewVoteClick(baseQuote, upVoteClickAction),
      calculateNewVoteClick(upVotedQuote, upVoteClickAction),
      calculateNewVoteClick(downVotedQuote, upVoteClickAction),
    ];
    expect(votesCases).toMatchSnapshot();
  });

  describe("downvote cases", () => {
    it("downvote cases", () => {
      const votesCases = [
        calculateNewVoteClick(baseQuote, downVoteClickAction),
        calculateNewVoteClick(upVotedQuote, downVoteClickAction),
        calculateNewVoteClick(downVotedQuote, downVoteClickAction),
      ];
      expect(votesCases).toMatchSnapshot();
    });
  });
});
