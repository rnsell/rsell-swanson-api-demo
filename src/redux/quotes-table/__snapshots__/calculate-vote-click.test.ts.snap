// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculate-vote-click downvote cases downvote cases 1`] = `
Array [
  Object {
    "downVoteSelected": true,
    "id": "7e9a20483bbbc00e4621614487a13194",
    "quoteText": "Go Beyond Plus Ultra",
    "upVoteSelected": false,
    "votes": 0,
  },
  Object {
    "downVoteSelected": true,
    "id": "7e9a20483bbbc00e4621614487a13194",
    "quoteText": "Go Beyond Plus Ultra",
    "upVoteSelected": false,
    "votes": -3,
  },
  Object {
    "downVoteSelected": false,
    "id": "7e9a20483bbbc00e4621614487a13194",
    "quoteText": "Go Beyond Plus Ultra",
    "upVoteSelected": false,
    "votes": 3,
  },
]
`;

exports[`calculate-vote-click upvote cases 1`] = `
Array [
  Object {
    "downVoteSelected": false,
    "id": "7e9a20483bbbc00e4621614487a13194",
    "quoteText": "Go Beyond Plus Ultra",
    "upVoteSelected": true,
    "votes": 1,
  },
  Object {
    "downVoteSelected": false,
    "id": "7e9a20483bbbc00e4621614487a13194",
    "quoteText": "Go Beyond Plus Ultra",
    "upVoteSelected": false,
    "votes": -1,
  },
  Object {
    "downVoteSelected": false,
    "id": "7e9a20483bbbc00e4621614487a13194",
    "quoteText": "Go Beyond Plus Ultra",
    "upVoteSelected": true,
    "votes": 2,
  },
]
`;
